# **Bangalore Tech Salary Decoder**

> **Live Project**

>**By Somyajeet Satapathy**




**Objective:** To analyze a dataset and cleaning the data, answering business questions, and generating salary insights using Python and Pandas.

**Task-1**

In [51]:
import pandas as pd
import numpy as np

In [52]:
df = pd.read_csv("bangalore_tech_salaries.csv")

In [53]:
df.head()  #Displaying first 5 rows

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite


In [54]:
df.tail() # Displaying last 5 rows

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
1010,BLR0824,Product Designer,2,17.6,12.4,Zoho,Mid-size,"React, SQL, ML, Deep Learning, Java",Jayanagar,Tier-2,2022,WFH
1011,BLR0792,Product Lead,4,₹40.2 LPA,26.6,HCL,MNC,"Excel, Java, Deep Learning, Python, JavaScript",Whitefield,Tier 2,2021,Work from Office
1012,BLR0209,BI Analyst,1,8.6 LPA,6.0 LPA,EdMint,Early-stage,"PyTorch, System Design, MongoDB, JIRA, PowerBI...",NaN,2,2024,Hybrid
1013,BLR0788,PM,2,27.6,17.4,Zomato,Unicorn,"Azure, Kafka, Excel",Bellandur,Tier-3,2023,Remote
1014,BLR0905,DA,4,₹20.3 LPA,15.4 LPA,Capgemini,MNC,"Deep Learning, JavaScript, NumPy",Marathahalli,Tier 2,2020,WFH


In [55]:
print("Dataset Shape:", df.shape)

Dataset Shape: (1015, 12)


In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB


In [57]:
df.dtypes  #for printing the data type

,0
Employee_ID,object
Role,object
years_exp,int64
Current_CTC,object
Previous_CTC,object
Company,object
company_TYPE,object
Skills,object
Location,object
Education_Tier,object


In [58]:
df.isnull().sum()  # Calculating missing values present in each column.

,0
Employee_ID,0
Role,0
years_exp,0
Current_CTC,0
Previous_CTC,200
Company,0
company_TYPE,0
Skills,27
Location,20
Education_Tier,0


In [59]:
print("Duplicate Rows:", df.duplicated().sum())   # Counting duplicate records

Duplicate Rows: 15


In [60]:
df.describe()

,years_exp,Joining_Year
count,1015.000000,1015.000000
mean,3.429557,2022.698522
std,2.558721,1.308351
min,0.000000,2020.000000
25%,1.000000,2022.000000
50%,3.000000,2023.000000
75%,5.000000,2024.000000
max,12.000000,2024.000000


**Task-2**

In [61]:
#AI asissted
#Renaming column names to snake consistency
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

print(df.columns)

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')


In [62]:
print("Role Variants:\n")     #Chcking for unique values
print(df["role"].unique())
print("\nCompany Type Variants:\n")
print(df["company_type"].unique())
print("\nEducation Tier Variants:\n")
print(df["education_tier"].unique())
print("\nWork Mode Variants:\n")
print(df["work_mode"].unique())

Role Variants:

['DS' 'DevOps' 'SDE Backend' 'SDE Frontend' 'Product Lead'
 'Product Manager' 'Site Reliability Engineer' 'DA' 'UI/UX' 'SDE-Frontend'
 'UI Designer' 'ML Engineer' 'Frontend Dev' 'Data Scientist'
 'Full Stack Engineer' 'Designer' 'Product Designer' 'BA'
 'Frontend Engineer' 'BI Analyst' 'Backend Dev' 'Sr PM' 'FE' 'SDE-Backend'
 'Data Analyst' 'Business Analyst' 'Backend Developer'
 'Business Systems Analyst' 'Data Science Engineer' 'FullStack' 'BE'
 'Frontend Developer' 'Backend Engineer' 'UX Designer' 'Infra Engineer'
 'PM' 'SDE FS' 'DevOps Engineer' 'Analytics Engineer' 'Fullstack Dev'
 'SDE Full-Stack' 'SRE']

Company Type Variants:

['Unicorn' 'MNC' 'Mid-size' 'early-stage' 'Early-stage' 'EARLY-STAGE'
 'mnc' 'unicorn' 'MID-SIZE' 'mid-size' 'UNICORN']

Education Tier Variants:

['Tier-1' '2' 'Tier-2' 'Tier 1' 'Tier 2' 'T3' 'Tier-3' 'T2' '1' 'T1'
 'Tier 3' '3']

Work Mode Variants:

['Remote' 'Hybrid' 'Work from Office' 'Onsite' 'Hybrid (3 days)' 'WFH']


In [63]:
#AI Assisted: to help create a mapping dictionary
#Standardizing different role names into common canonical roles.
role_mapping = {
    "SDE Backend": "SDE Backend",
    "SDE-Backend": "SDE Backend",
    "Backend Dev": "SDE Backend",
    "Backend Developer": "SDE Backend",
    "Backend Engineer": "SDE Backend",
    "BE": "SDE Backend",
    "SDE Frontend": "SDE Frontend",
    "SDE-Frontend": "SDE Frontend",
    "Frontend Dev": "SDE Frontend",
    "Frontend Developer": "SDE Frontend",
    "Frontend Engineer": "SDE Frontend",
    "FE": "SDE Frontend",
    "SDE Full-Stack": "SDE Full-Stack",
    "Full Stack Engineer": "SDE Full-Stack",
    "Fullstack Dev": "SDE Full-Stack",
    "FullStack": "SDE Full-Stack",
    "SDE FS": "SDE Full-Stack",
    "DS": "Data Scientist",
    "Data Scientist": "Data Scientist",
    "Data Science Engineer": "Data Scientist",
    "DA": "Data Analyst",
    "Data Analyst": "Data Analyst",
    "Analytics Engineer": "Data Analyst",
    "BI Analyst": "Data Analyst",
    "Business Analyst": "Business Analyst",
    "Business Systems Analyst": "Business Analyst",
    "BA": "Business Analyst",
    "Product Manager": "Product Manager",
    "PM": "Product Manager",
    "Product Lead": "Product Lead",
    "Sr PM": "Product Lead",
    "ML Engineer": "ML Engineer",
    "DevOps": "DevOps Engineer",
    "DevOps Engineer": "DevOps Engineer",
    "Site Reliability Engineer": "Site Reliability Engineer",
    "SRE": "Site Reliability Engineer",
    "Infra Engineer": "Site Reliability Engineer",
    "UI Designer": "UI/UX Designer",
    "UX Designer": "UI/UX Designer",
    "Designer": "UI/UX Designer",
    "UI/UX": "UI/UX Designer",
    "Product Designer": "UI/UX Designer"
}

df["role_clean"] = df["role"].replace(role_mapping)
print(df["role_clean"].value_counts())

role_clean
SDE Backend                  116
UI/UX Designer               116
Business Analyst             112
SDE Frontend                 105
Data Analyst                 101
SDE Full-Stack                97
Data Scientist                96
Site Reliability Engineer     81
Product Lead                  58
Product Manager               55
DevOps Engineer               46
ML Engineer                   32
Name: count, dtype: int64


In [64]:
company_mapping = {    # Convert different company type spellings into a standard format.
    "Unicorn": "Unicorn",
    "unicorn": "Unicorn",
    "UNICORN": "Unicorn",
    "MNC": "MNC",
    "mnc": "MNC",
    "Mid-size": "Mid-size",
    "mid-size": "Mid-size",
    "MID-SIZE": "Mid-size",
    "Early-stage": "Early-stage",
    "early-stage": "Early-stage",
    "EARLY-STAGE": "Early-stage"
}
df["company_type_clean"] = df["company_type"].replace(company_mapping)
print(df["company_type_clean"].value_counts())

company_type_clean
MNC            348
Mid-size       261
Unicorn        207
Early-stage    199
Name: count, dtype: int64


In [65]:
# Standardizing all education tier formats into three common categories.
education_mapping = {
    "Tier 1": "Tier 1",
    "Tier-1": "Tier 1",
    "T1": "Tier 1",
    "1": "Tier 1",
    "Tier 2": "Tier 2",
    "Tier-2": "Tier 2",
    "T2": "Tier 2",
    "2": "Tier 2",
    "Tier 3": "Tier 3",
    "Tier-3": "Tier 3",
    "T3": "Tier 3",
    "3": "Tier 3"
}
df["education_tier_clean"] = df["education_tier"].replace(education_mapping)
print(df["education_tier_clean"].value_counts())

education_tier_clean
Tier 2    510
Tier 3    309
Tier 1    196
Name: count, dtype: int64


In [66]:
# Placing in Emote, hybrid and Onsite

work_mode_mapping = {
    "Remote": "Remote",
    "WFH": "Remote",
    "Hybrid": "Hybrid",
    "Hybrid (3 days)": "Hybrid",
    "Onsite": "Onsite",
    "Work from Office": "Onsite"
}
df["work_mode_clean"] = df["work_mode"].replace(work_mode_mapping)
print(df["work_mode_clean"].value_counts())

work_mode_clean
Hybrid    347
Remote    340
Onsite    328
Name: count, dtype: int64


In [67]:
#AI assisted: for checking the handling process of ctc in single parsing
# Converting different salary formats into numeric LPA values.

def parse_ctc(ctc):
    if pd.isna(ctc):
        return np.nan
    ctc = str(ctc).strip()
    ctc = ctc.replace("₹", "")
    ctc = ctc.replace("LPA", "")
    ctc = ctc.replace("lpa", "")
    ctc = ctc.strip()
    if "," in ctc:        # Handle Indian currency format like 15,50,000
        number = ctc.replace(",", "")
        if number.isdigit():
            number = float(number)
            if number > 1000:
                return number / 100000

        return pd.to_numeric(number, errors="coerce")
    return pd.to_numeric(ctc, errors="coerce")

In [68]:
df["current_ctc_lpa"] = df["current_ctc"].apply(parse_ctc)
df["previous_ctc_lpa"] = df["previous_ctc"].apply(parse_ctc)

In [69]:
df[["current_ctc", "current_ctc_lpa",
    "previous_ctc", "previous_ctc_lpa"]].head(10)

,current_ctc,current_ctc_lpa,previous_ctc,previous_ctc_lpa
0,49.4,49.4,32.4 LPA,32.4
1,9.7,9.7,NaN,NaN
2,"3,360,000",33.6,24.2 LPA,24.2
3,41.4 LPA,41.4,30.3,30.3
4,"3,640,000",36.4,30.6 LPA,30.6
5,32.1,32.1,22.0 LPA,22.0
6,24.6,24.6,19.3 LPA,19.3
7,21.7 LPA,21.7,17.7,17.7
8,21.0,21.0,NaN,NaN
9,26.9 LPA,26.9,18.9,18.9


In [70]:
df = df.drop_duplicates()
print("Remaining duplicate rows:", df.duplicated().sum())

Remaining duplicate rows: 0


In [71]:
df.isnull().sum()  #Checking missing value

,0
employee_id,0
role,0
years_exp,0
current_ctc,0
previous_ctc,194
company,0
company_type,0
skills,27
location,19
education_tier,0


In [72]:
df = df.dropna(subset=["current_ctc_lpa"])

In [73]:
# Previous CTC is intentionally missing for freshers, therefore, these missing values are left unchanged.
print("Previous CTC missing values:", df["previous_ctc_lpa"].isnull().sum())

Previous CTC missing values: 194


In [74]:
print(df.dtypes)  #Displaying Cleaned Datatypes

employee_id              object
role                     object
years_exp                 int64
current_ctc              object
previous_ctc             object
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
company_type_clean       object
education_tier_clean     object
work_mode_clean          object
current_ctc_lpa         float64
previous_ctc_lpa        float64
dtype: object


In [75]:
# Verifying cleaned columns.
print("\nRole Counts\n")
print(df["role_clean"].value_counts())
print("\nCompany Type Counts\n")
print(df["company_type_clean"].value_counts())
print("\nEducation Tier Counts\n")
print(df["education_tier_clean"].value_counts())


Role Counts

role_clean
SDE Backend                  114
UI/UX Designer               113
Business Analyst             108
SDE Frontend                 105
Data Analyst                 100
SDE Full-Stack                96
Data Scientist                94
Site Reliability Engineer     81
Product Lead                  57
Product Manager               55
DevOps Engineer               46
ML Engineer                   31
Name: count, dtype: int64

Company Type Counts

company_type_clean
MNC            342
Mid-size       257
Unicorn        207
Early-stage    194
Name: count, dtype: int64

Education Tier Counts

education_tier_clean
Tier 2    500
Tier 3    306
Tier 1    194
Name: count, dtype: int64


**Task-3**

**3.1: Calculating the median, mean, minimum, and maximum Current CTC (LPA) for each role**

In [76]:
role_salary = (
    df.groupby("role_clean")["current_ctc_lpa"]
    .agg(
        Median_CTC="median",
        Mean_CTC="mean",
        Minimum_CTC="min",
        Maximum_CTC="max"
    )
    .sort_values(by="Median_CTC", ascending=False)
)
print(role_salary)

                           Median_CTC   Mean_CTC  Minimum_CTC  Maximum_CTC
role_clean                                                                
Product Manager                 31.40  35.418182         13.8         80.1
Product Lead                    31.10  32.889473         10.8         65.5
ML Engineer                     27.50  30.335484         10.0         64.5
Data Scientist                  24.15  26.670213         10.9         75.6
Site Reliability Engineer       23.30  23.481481          8.8         55.0
SDE Full-Stack                  22.35  25.409375          8.9         71.7
SDE Frontend                    21.20  22.193333          6.7         84.4
SDE Backend                     21.10  23.284210          7.9         55.1
Business Analyst                19.90  21.873148          6.8         52.7
DevOps Engineer                 19.45  22.891304          9.7         60.3
UI/UX Designer                  18.90  20.968142          6.2         63.3
Data Analyst             

In [77]:
# Displaying highest and lowest paying roles
highest_role = role_salary.index[0]
lowest_role = role_salary.index[-1]
print("Highest Paying Role:", highest_role)
print("Median CTC:", role_salary.iloc[0]["Median_CTC"], "LPA")
print()
print("Lowest Paying Role:", lowest_role)
print("Median CTC:", role_salary.iloc[-1]["Median_CTC"], "LPA")

Highest Paying Role: Product Manager
Median CTC: 31.4 LPA

Lowest Paying Role: Data Analyst
Median CTC: 16.9 LPA


In [78]:
# Comparing mean and median salaries
role_salary["Difference"] = (role_salary["Mean_CTC"] - role_salary["Median_CTC"])
print(role_salary)

                           Median_CTC   Mean_CTC  Minimum_CTC  Maximum_CTC  \
role_clean                                                                   
Product Manager                 31.40  35.418182         13.8         80.1   
Product Lead                    31.10  32.889473         10.8         65.5   
ML Engineer                     27.50  30.335484         10.0         64.5   
Data Scientist                  24.15  26.670213         10.9         75.6   
Site Reliability Engineer       23.30  23.481481          8.8         55.0   
SDE Full-Stack                  22.35  25.409375          8.9         71.7   
SDE Frontend                    21.20  22.193333          6.7         84.4   
SDE Backend                     21.10  23.284210          7.9         55.1   
Business Analyst                19.90  21.873148          6.8         52.7   
DevOps Engineer                 19.45  22.891304          9.7         60.3   
UI/UX Designer                  18.90  20.968142          6.2   

**3.2: Analysing how median salary changes**

In [79]:
backend_df = df[df["role_clean"] == "SDE Backend"].copy()
backend_df["exp_band"] = pd.cut(
    backend_df["years_exp"],
    bins=[-1, 1, 3, 5, 100],
    labels=["0-1", "2-3", "4-5", "6+"]
)
backend_salary = (
    backend_df
    .groupby("exp_band")["current_ctc_lpa"]
    .median()
)
print(backend_salary)

exp_band
0-1    11.65
2-3    20.00
4-5    25.85
6+     40.40
Name: current_ctc_lpa, dtype: float64


/tmp/ipykernel_1361/682564892.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  .groupby("exp_band")["current_ctc_lpa"]


In [80]:
growth = backend_salary.pct_change() * 100
experience_table = pd.DataFrame({
    "Median_CTC": backend_salary,
    "Growth_%": growth
})
print(experience_table)

          Median_CTC   Growth_%
exp_band                       
0-1            11.65        NaN
2-3            20.00  71.673820
4-5            25.85  29.250000
6+             40.40  56.286267


**3.3: Comparing median salary to identify which skill provide highest salary**

In [81]:
sde_df = df[df["role_clean"].isin([
    "SDE Backend",
    "SDE Frontend",
    "SDE Full-Stack"
])].copy()
sde_df["skills"] = sde_df["skills"].fillna("")
skills = [
    "AWS",
    "ML",
    "System Design",
    "Kubernetes"
]
skill_results = []
for skill in skills:
    with_skill = sde_df[
        sde_df["skills"].str.contains(skill, case=False)
    ]["current_ctc_lpa"].median()
    without_skill = sde_df[
        ~sde_df["skills"].str.contains(skill, case=False)
    ]["current_ctc_lpa"].median()
    premium = ((with_skill - without_skill) / without_skill) * 100
    skill_results.append([
        skill,
        round(with_skill, 2),
        round(without_skill, 2),
        round(premium, 2)
    ])
skill_df = pd.DataFrame(
    skill_results,
    columns=[
        "Skill",
        "With Skill",
        "Without Skill",
        "Premium (%)"
    ]
)
print(skill_df)

           Skill  With Skill  Without Skill  Premium (%)
0            AWS       20.95           21.9        -4.34
1             ML       23.95           21.3        12.44
2  System Design       25.30           21.0        20.48
3     Kubernetes       23.20           21.5         7.91


**3.4: Calculating salary premium**

In [82]:
backend_company = df[df["role_clean"] == "SDE Backend"]
company_salary = (
    backend_company
    .groupby("company_type_clean")["current_ctc_lpa"]
    .median()
    .reindex(["Unicorn", "MNC", "Mid-size", "Early-stage"])
)

print(company_salary)

company_type_clean
Unicorn        27.35
MNC            20.45
Mid-size       19.50
Early-stage    18.60
Name: current_ctc_lpa, dtype: float64


In [83]:

unicorn_salary = company_salary["Unicorn"]
mnc_salary = company_salary["MNC"]
premium = ((unicorn_salary - mnc_salary) / mnc_salary) * 100
print(f"Unicorn Median CTC : {unicorn_salary:.2f} LPA")
print(f"MNC Median CTC      : {mnc_salary:.2f} LPA")
print(f"Premium             : {premium:.2f}%")

Unicorn Median CTC : 27.35 LPA
MNC Median CTC      : 20.45 LPA
Premium             : 33.74%


**3.5: identifying professionals earning more below their peers**

In [84]:
df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=[-1, 1, 3, 5, 100],
    labels=["0-1", "2-3", "4-5", "6+"]
)

In [85]:
group_size = (
    df.groupby(
        ["role_clean", "company_type_clean", "exp_band"]
    )["current_ctc_lpa"]
    .transform("count")
)
filtered_df = df[group_size >= 10].copy()

/tmp/ipykernel_1361/3232497210.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(


In [86]:
filtered_df["group_median"] = (    #Calculating median
    filtered_df.groupby(
        ["role_clean", "company_type_clean", "exp_band"]
    )["current_ctc_lpa"]
    .transform("median")
)

/tmp/ipykernel_1361/2855912234.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  filtered_df.groupby(


In [87]:
filtered_df["gap"] = (
    filtered_df["current_ctc_lpa"] -
    filtered_df["group_median"]
)

In [88]:
# Displaying the ten most underpaid professionals.
underpaid = (
    filtered_df.sort_values("gap")
    [
        [
            "employee_id",
            "role_clean",
            "company_type_clean",
            "years_exp",
            "current_ctc_lpa",
            "group_median",
            "gap"
        ]
    ].head(10)
)
print(underpaid)

    employee_id        role_clean company_type_clean  years_exp  \
904     BLR0925       SDE Backend                MNC          4   
250     BLR0072       SDE Backend                MNC          4   
249     BLR0733  Business Analyst                MNC          7   
386     BLR0453  Business Analyst                MNC          6   
711     BLR0451      SDE Frontend                MNC          4   
688     BLR0203       SDE Backend                MNC          4   
858     BLR0033      SDE Frontend                MNC          4   
938     BLR0525       SDE Backend                MNC          2   
466     BLR0375    Data Scientist                MNC          0   
94      BLR0328       SDE Backend                MNC          2   

     current_ctc_lpa  group_median  gap  
904             19.4          27.3 -7.9  
250             19.5          27.3 -7.8  
249             31.1          38.5 -7.4  
386             32.1          38.5 -6.4  
711             19.5          25.5 -6.0  
688       

**Task-4**

In [89]:
top5_underpaid = underpaid.head(5)  #top 5 underpaid professional

In [90]:
# AI Assisted: for formatting the code and checking the outputs
# Generating the final salary Intelligence Report.

print("=" * 60)
print("             BANGALORE TECH SALARY DECODER")
print("              Built by Somyajeet Satapathy")
print("=" * 60)
print(f"Dataset : {len(df)} Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")
print("\n" + "-" * 60)
print("MEDIAN CTC BY ROLE (LPA)")
print("-" * 60)

for role, row in role_salary.iterrows():
    bar = "█" * int(row["Median_CTC"] / 2)
    print(f"{role:<28} {bar:<20} {row['Median_CTC']:.2f}")
print("\n" + "-" * 60)
print("SDE BACKEND CTC BY EXPERIENCE")
print("-" * 60)
for band, row in experience_table.iterrows():
    growth = row["Growth_%"]
    if pd.isna(growth):
        print(f"{band:<10} : {row['Median_CTC']:.2f} LPA")
    else:
        print(f"{band:<10} : {row['Median_CTC']:.2f} LPA (+{growth:.2f}%)")
print("\n" + "-" * 60)
print("SKILL PREMIUM FOR SDEs")
print("-" * 60)
print(skill_df.to_string(index=False))
print("\n" + "-" * 60)
print("COMPANY TYPE PREMIUM (SDE Backend)")
print("-" * 60)
for company, salary in company_salary.items():
    if company == "Unicorn":
        print(f"{company:<15}{salary:.2f} LPA (Highest)")
    else:
        difference = ((salary - unicorn_salary) / unicorn_salary) * 100
        print(f"{company:<15}{salary:.2f} LPA ({difference:.2f}% vs Unicorn)")

print("\n" + "-" * 60)
print("TOP 5 MOST UNDERPAID PROFESSIONALS")
print("-" * 60)
for _, row in top5_underpaid.iterrows():
    print(
        f"{row['employee_id']}   "
        f"{row['role_clean']:<18} "
        f"{row['company_type_clean']:<10} "
        f"{row['years_exp']} yrs   "
        f"Gap: {row['gap']:.2f} LPA"
    )

print("=" * 60)

             BANGALORE TECH SALARY DECODER
              Built by Somyajeet Satapathy
Dataset : 1000 Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot

------------------------------------------------------------
MEDIAN CTC BY ROLE (LPA)
------------------------------------------------------------
Product Manager              ███████████████      31.40
Product Lead                 ███████████████      31.10
ML Engineer                  █████████████        27.50
Data Scientist               ████████████         24.15
Site Reliability Engineer    ███████████          23.30
SDE Full-Stack               ███████████          22.35
SDE Frontend                 ██████████           21.20
SDE Backend                  ██████████           21.10
Business Analyst             █████████            19.90
DevOps Engineer              █████████            19.45
UI/UX Designer               █████████            18.90
Data Analyst                 ████████             16.90

--------------

### **Task-5**

### Insights

### 1.
Product Managers have the highest median salary (**31.40 LPA**), while Data Analysts have the lowest (**16.90 LPA**). Students aiming for higher compensation should consider developing product management or leadership skills.

### 2.
For SDE Backend professionals, moving from **0–1 years** to **2–3 years** of experience results in the largest salary increase (**71.67%**). Building strong technical skills helps in growth of career.

### 3.
Among SDE professionals, **System Design** provides the highest salary premium (**20.48%**). Learning System Design can improve job opportunity.